### PR TITLE
Glob rewrite to use minimatch

### LINKED ISSUE #9: Unconventional globbing behavior
This project includes a pattern matching engine for selecting files. It's referred to as "globbing" because it's modeled on the "globing" syntax implemented in many other programming environments. While using the software, I found a number of cases where the behavior does not match convention. I'll summarize here and demonstrate convention with examples from [GNU Bash](https://www.gnu.org/software/bash/) and [Node.js minimatch](https://www.npmjs.com/package/minimatch).

In this project's implementation, the start character (`*`) matches one or more characters (there's no test for this behavior).

Conventionally, it matches zero or more characters:

    $ touch hello h_ello
    $ ls h*ello
    hello  h_ello

    > minimatch('hello', 'h*ello')
    true
    > minimatch('h_ello', 'h*ello')
    true

In this project's implementation, the comma character behaves as a logical "or" wherever it appears:

    assertGlobTrue('a,b', 'a');
    assertGlobTrue('a,b', 'b');
    assertGlobTrue('{a,b}', 'a');
    assertGlobTrue('{a,b}', 'b');

Conventionally, `,` matches only the literal comma character, with the "option" behavior limited to applications within braces:

    $ touch a b a,b
    $ ls a,b
    a,b
    $ ls {a,b}
    a  b

    > minimatch('a', 'a,b'));
    false
    > minimatch('b', 'a,b'));
    false
    > minimatch('a,b', 'a,b'));
    true
    > minimatch('a', '{a,b}'));
    true
    > minimatch('b', '{a,b}'));
    true
    > minimatch('a,b', '{a,b}'));
    false

In this project's implementation, the solidus is interpreted as both a literal forward slash character and a delimiter for partial matching:

    assertGlobTrue('a/b', 'a');
    assertGlobTrue('a/b', 'a/');
    assertGlobTrue('a/b', 'a/b');

Conventionally, `/` only functions as a literal forward slash:

    $ mkdir a; touch a/b
    $ ls a/b
    a/b

    > minimatch('a', 'a/b')
    false
    > minimatch('a/', 'a/b')
    false
    > minimatch('a/b', 'a/b')
    true

We could probably extend the current implementation to address these cases, but that will further complicate a component of the system which is only orthogonally related to the problem domain. I'm also not convinced that this report is exhaustive, so there may be still other unconventional behaviors present. In the interest of reducing complexity and minimizing surprise, I'd like to revisit my earlier suggestion to use an external project.

### PR DESCRIPTION
**Pull request by gnarf**
Fixes #9 

Breaking change: previously the "files" parameter would treat a `,` as a glob separator, in order to continue treating `,` as a separator, it will now need to be wrapped in `{}` like a glob would expect.  The program still supports passing multiple globs as additional parameters:

An example - an old command from my bash history:
`node aria-at-automation-harness/bin/host.js run-plan --plan-workingdir aria-at/build/tests/menu-button-actions-active-descendant "reference/**,scripts/**,test-17*-nvda.*"`

will now be

`node aria-at-automation-harness/bin/host.js run-plan --plan-workingdir aria-at/build/tests/menu-button-actions-active-descendant "reference/**" "scripts/**" "test-17*-nvda.*"`

or

`node aria-at-automation-harness/bin/host.js run-plan --plan-workingdir aria-at/build/tests/menu-button-actions-active-descendant "{reference/**,scripts/**,test-17*-nvda.*}"`

## May interfere with the automation until these PR are landed:
* [x] w3c/aria-at-app#916
* [x] bocoup/aria-at-gh-actions-helper#5

### CODE REVIEW

File: src/host/tests/snapshots/plan-from.js.md
Comment (gnarf) [inline src/host/tests/snapshots/plan-from.js.md:4843]: This is correct, the default test match pattern is `{,**/}test*` so `tests/1.test.json` and `tests/2.test.json` should never have been matched here (all further instances of this are this change)

File: src/shared/tests/file-glob.js
Comment (gnarf) [inline src/shared/tests/file-glob.js:36]: I still need to go through and update this test file to indicate partial (possible parent directory) matches.

File: package.json
Comment (gnarf) [inline package.json:1]: really unsure why `npm install --save minimatch` generated this big of a diff

File: src/host/tests/plan-from.js
Comment (gnarf) [inline src/host/tests/plan-from.js:9]: note to self: revert this file - no actual code changes made here

File: package.json
Comment (jugglinmike) [reply package.json:1]: It switched the file to Windows line endings. Any chance you ran the command on a Windows machine?

Comment (gnarf) [reply package.json:1]: I'll try to get the cr/lf noise reduced on this json before we land

**Comment by gnarf** [comment]
minimatch doesn't match `\` for path separator while on linux - looking for a way to force support i found https://github.com/isaacs/minimatch/blob/main/src/index.ts#L1133 - seems I can pass a `{platform: 'win32'}` to the options in the constructor (but that adds a whole new param to compileGlob) and I don't think we want to always be pretending to be windows in production (just for testing)

@jugglinmike got an opinion on how to handle this one? Couple of options that I see (and open to others):

* Rig up the win32 flag in `compileGlob` as a second parameter
* Rig up some other flag the suite can set `compileGlob.options = {}` or something?
* Just delete the tests that test backslashes, do we really need a test suite testing minimatch?

**Comment by jugglinmike** [comment]
I don't necessarily object to testing some of Minimatch's behavior since tests like that will guard against regressions should we ever decide to swap libraries. At the same time, the tests for backslash as a path separator are inappropriate for POSIX environments.

How about [skipping them](https://github.com/avajs/ava/blob/2e0c2b1cef779e1c092eb60f0a9558bb9cf4c848/docs/01-writing-tests.md#skipping-tests) outside of Windows?

**Comment by gnarf** [comment]
> How about [skipping them](https://github.com/avajs/ava/blob/2e0c2b1cef779e1c092eb60f0a9558bb9cf4c848/docs/01-writing-tests.md#skipping-tests) outside of Windows?

Since theses were all individual asserts, and not a whole test, I switched to `assertGlobTrueWin32` which tests for true result when in win32 and tests for false result in other environments.

File: src/host/tests/snapshots/plan-from.js.md
Comment (jugglinmike) [reply src/host/tests/snapshots/plan-from.js.md:4843]: It's great to learn that this change is correcting previously-unreported errors! And while I fully support keeping these tests by correcting their expectations, *only* doing that feels like a missed opportunity. What do you think about adding new tests which correctly reflect the original intent (e.g. with a pattern like `{,**/*}test*`)?

File: src/shared/file-glob.js
Comment (jugglinmike) [inline src/shared/file-glob.js:32]: [The second parameter to the `match` method is undocumented.](https://www.npmjs.com/package/minimatch#methods), so its use would make this project vulnerable to regressions in future releases of minimatch. I think we'll have to forgo the memoization and construct a new object for each "match" calculation.

Separately, it wouldn't hurt to petition the maintainers of minimatch to commit to that API, but I wouldn't gate this patch on the resolution of such a proposal (though if you do file an issue, referencing it in a comment would certainly be helpful).

File: src/shared/file-record.js
Comment (jugglinmike) [inline src/shared/file-record.js:143]: I appreciate the care you're taking here, but this seems like a failure which could impact the validity of the results. If it occurred in the GitHub Actions environment, logging a warning would only help us verify that something went wrong after reporting faulty results--it wouldn't stop them from being reported at all. What do you say about crashing for this case?

File: src/host/tests/snapshots/plan-from.js.md
Comment (gnarf) [reply src/host/tests/snapshots/plan-from.js.md:4843]: This test already exists.  Specifically `for (const testPattern of [undefined, '*.json', 'tests/*'])` - these corrected results are only in the `undefined` testPattern - the `*.json` and `tests/*` both still match this

File: src/shared/file-glob.js
Comment (gnarf) [reply src/shared/file-glob.js:32]: Nice catch - I switched to caching 2 matchers - wastes a little bit of time in the test suite, but the main use for this it will still save a bunch of re-compiling the glob

File: src/host/tests/snapshots/plan-from.js.md
Comment (gnarf) [reply src/host/tests/snapshots/plan-from.js.md:4843]: Correction: only the `tests/*` pattern finds the file (which is correct) - I incorrectly stated `*.json` also was (which would be wrong)

File: src/shared/file-record.js
Comment (jugglinmike) [reply src/shared/file-record.js]: Cool

File: src/shared/file-glob.js
Comment (jugglinmike) [reply src/shared/file-glob.js]: Great idea!

**Review by jugglinmike — APPROVED** [review]
Thanks, Corey!

**Pull request merged**
